enarx-keepmgr/src/drivers/mod.rs

```diff
@@ -0,0 +1,111 @@
+//! # Overview
+//!
+//! The traits contained in this module constitute a state machine for
+//! launching and managing a keep across multiple technologies. Launching
+//! a keep goes through multiple phases, in this order:
+//!
+//!   1. Driver
+//!
+//!   2. Setup
+//!     a. Shim Setup
+//!     b. Keep Setup
+//!
+//!   3. Build
+//!     a. Shim Build
+//!     b. Keep Build
+//!
+//!   4. Enter
+//!
+//! # Details
+//! ## The Driver Phase
+//!
+//! First, the keep manager must find an appropriate driver for the system.
+//! Then, the keep manager must locate the keep runtime and the driver-supplied
+//! shim for use in the next phase. When the keep manager is ready to launch
+//! a keep, it uses `Driver::make()` to begin the setup phase.
+//!
+//! ## The Setup Phase
+//!
+//! The purpose of the setup phase is to give the driver a view of both the
+//! keep runtime and shim before creating the actual keep. This is useful for
+//! technologies like SGX where we need to hash and sign the binary before
+//! creating the keep.
+//!
+//! The keep manager will load the driver-supplied shim first, followed by
+//! keep runtime. Upon completion, we begin the build phase.
+//!
+//! ## The Build Phase
+//!
+//! The purpose of the building phase is to actually create a full keep
+//! with all the pages actually loaded. The workflow is similar to the
+//! setup phase.
+//!
+//! The keep manager will load the driver-supplied shim first, followed by
+//! keep runtime. Upon completion, we begin the enter phase.
+//!
+//! ## The Enter Phase
+//!
+//! The enter phase can be considered the main loop of the application.
+//! When the keep manager enters the keep, the keep will run until an event
+//! occurs. Upon an event, the keep manager will handle the event, potentially
+//! collecting the keep's required data, and re-entering the keep.
+
```

Comment:
Is it correct that attestation would happen after this? (The Keep needs to be running in order to issue a Report about itself, at least for SGX.)

Reply:
For `SGX`, yes. For `SEV`, no.